### ISSUE TITLE
[Ch 20-21][Notebook][Major] SFT 챕터를 Ch 27 로 지칭 — 실제 SFT 는 Ch 28 (2개 챕터 5곳)

### ISSUE THREAD

**Issue by yum-yeom** [yoon-gu/neuqes-101#121]
> 📓 **[Ch 20 Colab](https://colab.research.google.com/github/yoon-gu/neuqes-101/blob/master/20_en_bert_pretrain/20_en_bert_pretrain.ipynb)** · **[Ch 21 Colab](https://colab.research.google.com/github/yoon-gu/neuqes-101/blob/master/21_en_bert_classify/21_en_bert_classify.ipynb)** · **[Ch 28 — 실제 SFT 챕터](https://colab.research.google.com/github/yoon-gu/neuqes-101/blob/master/28_sft/28_sft.ipynb)**

## 대상

- **산출물**: `20_en_bert_pretrain.ipynb` 셀 22 (1곳) · `21_en_bert_classify.ipynb` 셀 21 (4곳). 전파: `pages/`, `book/chapters/`
- **기준 커밋**: `bb9af78` (`fix/ch20-mask-top5-interpretation`) · `c16215e` (`fix/ch21-review-2026-09`)
- **환경**: 정적 검토. Colab 재실행 미수행 (마크다운 전용)

## 실제 결과

두 챕터가 SFT / Instruction Tuning 을 **Ch 27** 로 지칭하는데, 실제 SFT 는 **Ch 28**(`28_sft/`) 이고 Ch 27 은 KoGPT2 Continual Pretraining 입니다.

**노트북 원문 — 5곳**

```markdown
① Ch 20 셀 22
> … SFT (Ch 27) 는 *prompt 만 -100, 답변만 학습*.

② Ch 21 셀 21 — 표 행
| **SFT / Instruction Tuning** (Ch 27) | **prompt 부분 = `-100`** | … |

③ Ch 21 셀 21 — 인용문
> … Phase 4 (특히 Ch 27 SFT, `SFTTrainer` 의 `response-only mask` 옵션) 에서 …

④ Ch 21 셀 21 — 다음 문단
… 기억해 두면, Ch 27 의 *왜 모델이 instruction 을 따라가게 되는가* 가 …

⑤ Ch 21 셀 21 — 파인튜닝 비교표
| **GPT 파인튜닝 = SFT** (Ch 25, Ch 27) |
```

**같은 셀의 다른 참조는 정확합니다** — `| **GPT CausalLM 사전학습** (Ch 24-26) |`. **SFT 참조만 한 칸씩 앞으로 밀려 있습니다.**

#83(sklearn 챕터 한 칸 밀림)과 같은 계열 — 챕터 번호 체계가 바뀔 때 전방 참조가 따라가지 못한 패턴이라 한 곳만 고치면 다시 어긋납니다.

## 기대 결과

SFT 를 지칭하는 모든 자리가 Ch 28 을 가리키고, Ch 27(Continual Pretraining)로 남아야 할 자리와 구분될 것.

## 수정 제안

**①~④ 는 번호 치환으로 끝납니다.**

```diff
- SFT (Ch 27) 는 *prompt 만 -100, 답변만 학습*.                                (Ch 20 셀 22)
+ SFT (Ch 28) 는 *prompt 만 -100, 답변만 학습*.
- | **SFT / Instruction Tuning** (Ch 27) | …                                  (Ch 21 셀 21)
+ | **SFT / Instruction Tuning** (Ch 28) | …
- … Phase 4 (특히 Ch 27 SFT, `SFTTrainer` 의 `response-only mask` 옵션) …
+ … Phase 4 (특히 Ch 28 SFT, `SFTTrainer` 의 `response-only mask` 옵션) …
- … 기억해 두면, Ch 27 의 *왜 모델이 instruction 을 따라가게 되는가* 가 …
+ … 기억해 두면, Ch 28 의 *왜 모델이 instruction 을 따라가게 되는가* 가 …
```

**⑤ 는 원저자 판단이 필요합니다.** `27 → 28` 치환만으로는 부족합니다 — Ch 25 는 `gpt2` Continual Pretraining 이라 "SFT 예시" 로 여전히 부정확합니다.

```diff
案 1 — SFT 계열만 가리키려던 것이라면
- | **GPT 파인튜닝 = SFT** (Ch 25, Ch 27) |
+ | **GPT 파인튜닝 = SFT** (Ch 28) |

案 2 — continual pretraining 을 포함한 넓은 의미였다면
+ | **GPT 시대의 추가 학습** (continual pretraining Ch 25·27, SFT Ch 28) |

문단 끝 — 어느 案이든 공통
- … *같은 모델* 이 다른 일을 합니다. Ch 27 에서 직접 확인.
+ … *같은 모델* 이 다른 일을 합니다. Ch 28 에서 직접 확인.
```

전 챕터 잔여 확인: `grep -rn 'Ch 27' --include='*.ipynb' --include='*.md' . | grep -i 'sft\|instruction'`

마크다운 전용이라 Colab 재실행 불필요 (#118 선례). 수정 후 §8 원칙대로 두 챕터 WikiDocs·EPUB 재변환·재빌드, `21_en_bert_classify/README.md` 등 변환 경로 밖 파일 별도 확인.

## 재현 절차

```bash
ls -d 27_* 28_*
python3 -c "
import json
for p in ['20_en_bert_pretrain/20_en_bert_pretrain.ipynb',
          '21_en_bert_classify/21_en_bert_classify.ipynb']:
    nb=json.load(open(p))
    for i,c in enumerate(nb['cells']):
        for l in ''.join(c['source']).split('\n'):
            if 'Ch 27' in l: print(f'{p} [{i}] {l.strip()[:140]}')
"
```

## 심각도

**Major** — Phase 4 로 넘어가는 다리로 세운 설명이 존재하지 않는 대응 관계를 가리킵니다. 학습자가 Ch 27 을 열면 SFT 가 아니라 Continual Pretraining 이 나옵니다. 실행은 막지 않습니다.

## 연결

#44 · #45 (챕터 검수) · #83 · #112 · #115 (같은 계열) · #118 (마크다운 전용 수정 선례)

**Comment by yum-yeom** [comment]
수정 완료 확인, 이슈 닫습니다.

- **Ch 20 쪽 (①)**: PR #124 로 수정·머지 확인 — master 의 `20_en_bert_pretrain/` 에 SFT 문맥 "Ch 27" 잔존 0.
- **Ch 21 쪽 (②③④⑤)**: `ch-21/review-2026-09` @ [`ccca546`](https://github.com/yoon-gu/neuqes-101/commit/ccca546) 으로 5곳 정정. ⑤ 는 **案 1** 로 — 비교표의 행 서술(라벨=모범 답안 시퀀스, 학습되는 것=행동 정렬)이 전부 SFT 전용이라 열 제목의 챕터 번호만 바로잡았습니다: `(Ch 25, Ch 27)` → `(Ch 28)`.
- 전 챕터 재스캔(이슈 본문의 grep) 결과 SFT 문맥의 Ch 27 오지칭 잔여 0건. `pages/`·tex 는 Ch 21 재변환 시 함께 반영됩니다.
